core/index/index_writer.cpp

```diff
@@ -1464,11 +1464,14 @@ bool index_writer::consolidate(
     if (candidates.empty()) {
       return;
     }
-
-    SCOPED_LOCK(consolidation_lock_);
+    SCOPED_LOCK_NAMED(consolidation_lock_, cleanup_lock);
+    decltype(flush_context::segment_mask_) cached_mask;
     for (const auto* candidate : candidates) {
+      cached_mask.insert(*candidate);
       consolidating_segments_.erase(candidate);
     }
+    cleanup_lock.unlock();
+    cached_readers_.purge(cached_mask);
```

Comment:
not sure if it's a good idea to cleanup cache right here given there might be ongoing removals